controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/filter/SignatureFilter.java

```diff
@@ -47,23 +55,40 @@ public SignatureFilter(Controller controller) {
             && request.getHeader("X-Authorization") != null)
             try {
                 ApplicationId id = ApplicationId.fromSerializedForm(request.getHeader("X-Key-Id"));
-                boolean verified = controller.applications().getApplication(TenantAndApplicationId.from(id)).stream()
-                                             .flatMap(application -> application.pemDeployKeys().stream())
-                                             .map(key -> new RequestVerifier(key, controller.clock()))
-                                             .anyMatch(verifier -> verifier.verify(Method.valueOf(request.getMethod()),
-                                                                                   request.getUri(),
-                                                                                   request.getHeader("X-Timestamp"),
-                                                                                   request.getHeader("X-Content-Hash"),
-                                                                                   request.getHeader("X-Authorization")));
+                SecurityContext securityContext = null;
+                if (request.getHeader("X-Key") != null) {
+                    PublicKey key = KeyUtils.fromPemEncodedPublicKey(new String(Base64.getDecoder().decode(request.getHeader("X-Key")), UTF_8));
+                    if (keyVerifies(key, request)) {
+                        Principal principal = null;
+                        Set<Role> roles = new HashSet<>();
+                        Optional <Application> application = controller.applications().getApplication(TenantAndApplicationId.from(id));
+                        if (application.isPresent() && application.get().deployKeys().contains(key)) {
+                            principal = new SimplePrincipal("headless@" + id.tenant() + "." + id.application());
+                            roles.add(Role.reader(id.tenant()));
+                            roles.add(Role.headless(id.tenant(), id.application()));
+                        }
+
+                        Optional<CloudTenant> tenant = controller.tenants().get(id.tenant())
+                                                                 .filter(CloudTenant.class::isInstance)
+                                                                 .map(CloudTenant.class::cast);
+                        if (tenant.isPresent() && tenant.get().developerKeys().containsKey(key)) {
+                            principal = tenant.get().developerKeys().get(key); // Precedence over headless user.
+                            roles.add(Role.reader(id.tenant()));
+                            roles.add(Role.developer(id.tenant()));
+                        }
+
+                        if (principal != null)
+                            securityContext = new SecurityContext(principal, roles);
+                    }
+                }
```

Comment:
Consider moving this branch to a separate method to make the separation between developer and deploy keys more clear.

Reply:
Current code is better, but will consider this when I remove outdated code paths again in a couple of weeks.